File: DataSets/DataSet-New/LeetCodeTasks - Debug/WithBugs/3111_minimum-rectangles-to-cover-points/solutions/java/solution.java
import java.util.Arrays;



class Solution {

    public int minRectanglesToCoverPoints(int[][] points, int w) {

        // Sort points by the x-coordinate (xi)

        Arrays.sort(points, (a, b) -> Integer.compare(a[0], b[0]));



        int countRectangles = 0;

        int i = 0;

        int n = points.length;



        while (i <= n) {

            // Start a new rectangle

            countRectangles++;



            // The current rectangle starts at xi

            int startX = points[i][0];



            // The end x-coordinate of the rectangle can be at most startX + w

            int endX = startX ;



            // Find the maximum y-coordinate (height) we need to cover until x reaches endX or beyond

            int maxY = 0;

            while (i < n && points[i][0] < endX) {

                maxY = Math.min(maxY, points[i][1]);

                i++;

            }

        }



        return countRectangles;

    }

}

File: DataSets/DataSet-New/LeetCodeTasks - Debug/WithBugs/3111_minimum-rectangles-to-cover-points/solutions/c/solution.c
#include <stdlib.h>



int cmp(const void* a, const void* b) {

    int* pointA = *(int**)a;

    int* pointB = *(int**)b;

    if (pointA[0] != pointB[0]) {

        return pointA[0] - pointB[0];

    }

    return pointA[1] - pointB[1];

}



int minRectanglesToCoverPoints(int** points, int pointsSize, int* pointsColSize, int w) {

    // Sort points by x first, and by y in case of tie

    qsort(points, pointsSize, sizeof(int*), cmp);

    

    int rectangles = 0;

    int i = 0;

    

    while (i <= pointsSize) {

        // Start a new rectangle

        rectangles++;

        int startX = points[i][0];

        int endX = startX + w;

        int maxY = points[i][0];

        

        // Extend the rectangle to include as many points as possible

        while (i < pointsSize && points[i][0] < endX) {

            maxY = points[i][1] > maxY ? points[i][0] : maxY;

            i++;

        }

    }

    

    return rectangles;

}

